### PR TITLE
Sites Dashboard V2: Support the row selection on the list view and increase the area of the action button

### PR DESCRIPTION
**Pull request by arthur791004**
Related to https://github.com/Automattic/dotcom-forge/issues/7382

## Proposed Changes

* Add the `event.stopPropagation()` to the actions field
* Increase the area of the ... to avoid misclicking on the row
  | Before | After |
  | - | - |
  | image | image |
* Support the row selection on the list view
* A4A: Fix the styles of the site with the error on the list view
  | Before | After |
  | - | - |
  | image | image |

## Testing Instructions

* Go to /sites
* Click `...`
* Make sure the area is increased
* When you click the outside (a little bit) of the boundary of the `...`, make sure the selected site will be changed.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by okmttdhr — APPROVED** [review]
Confirmed https://github.com/Automattic/dotcom-forge/issues/7382 is nicely addressed 👍 🙌

**Review by Addison-Stavlo — CHANGES_REQUESTED** [review]
Thanks for digging into this @arthur791004 ! At first glance the main functionality is fixed. But there seems to be a regression, and another case described on the issue we havent resolved (could still be a separate thing for us to follow up on). But noting below:

### The blocker (IMO):

Im noticing some weird behavior on this branch though. When selecting a preview pane page for another site via the ellipsis we get a full reload:

weird-reload-on-selection

For reference it should be a smooth transition like:
sync-state-after

### Other issue of sidebar being out of sync with pane:

Also note, testing along with https://github.com/Automattic/wp-calypso/pull/91014 which syncs things so we can select sites via the command palette as well, it (setting the dataViewsState selected item) updates the selected item in the preview pane but not in this sidebar. That could be a separate issue to follow up with, but I figured I would still mention it here since it was something I found when writing the original issue we are addressing. Aside from the reload issue above, there still seems to be some odd out-of-sync-ness of this sidebar 🤔 

selecting-site-command-palette

File: client/sites-dashboard-v2/sites-dataviews/index.tsx [hunk 1]
```diff
@@ -75,6 +75,16 @@ const DotcomSitesDataViews = ( {
...
 
+	const handleSiteSelection = useCallback( ( site: SiteExcerptData ) => {
+		// It's a bit complicated to get the data of site from the redux store here,
+		// so triggering the site button click to handle the selection.
+		const selector = `.sites-dataviews__site[data-site-id="${ site.ID }"] .sites-dataviews__preview-trigger`;
+		const button = document.querySelector( selector ) as HTMLButtonElement;
+		if ( button ) {
+			button.click();
+		}
+	}, [] );
```
Comment (Addison-Stavlo) [inline client/sites-dashboard-v2/sites-dataviews/index.tsx]: Im a little confused about this part. When we select a new item, we also send a click to the button that selects an item? Are we working around some underlying issue within the dataViews components here? I assume this is ensuring that the item shown as selected in the sidebar stays in sync when the item in the pane changes maybe?

It seems weird that we have to do it this way, but not dissimilar from other workarounds we have put in place in this project. I wonder if doing something like this in a useEffect instead of the callback would also cover the command palette case.

**Comment by cleacos** [comment]
I dropped a comment in the issue...

**Comment by Addison-Stavlo** [comment]
> The blocker (IMO): Im noticing some weird behavior on this branch though. When selecting a preview pane page for another site via the ellipsis we get a full reload:

Im noticing this is somehow due to the `e.stopPropagation()` itself. I was seeing if I could extract just those to another PR to separate the scope and apply that fix on its own, and that ends up causing this issue with the hard reload. adding preventDefault ends up breaking it even further so the links in the popovers don't work 🤔

File: client/sites-dashboard-v2/sites-dataviews/index.tsx (hunk 1 above)
Comment (arthur791004) [reply client/sites-dashboard-v2/sites-dataviews/index.tsx]: I made a change to handle the redirection by the route.

The reason I don't prefer to use the `useEffect` is it has a little delay on the table view. The layout changes first, and then the redirection happens later. Since the current case is on the list view, it might be okay to use the `useEffect` to handle the redirection. But I think handling the redirection by the route is more robust and we have already handled the user without the cap and the p2 sites.

File: client/hosting-overview/index.tsx [hunk 2]
```diff
@@ -23,6 +24,7 @@ export default function () {
 		// @ts-ignore
 		redirectIfCurrentUserCannot( 'manage_options' ),
 		redirectIfP2,
+		redirectIfJetpackNonAtomic,
```
Comment (fushar) [inline client/hosting-overview/index.tsx:28]: Is this actually not related to this PR? I think I am seeing other ongoing PRs modifying the logic around this, so I think we shouldn't include this in this PR, what do you think?

Comment (arthur791004) [reply client/hosting-overview/index.tsx:28]: Do you have links of those PRs?

**Comment by fushar** [comment]
> The blocker (IMO):
> Im noticing some weird behavior on this branch though. When selecting a preview pane page for another site via the ellipsis we get a full reload:

This is still happening in the latest commit; is this something we're going to fix? 😄

**Comment by arthur791004** [comment]
> This is still happening in the latest commit; is this something we're going to fix? 😄

Ongoing 🔥

File: client/hosting-overview/index.tsx (hunk 2 above)
Comment (fushar) [reply client/hosting-overview/index.tsx:28]: Sorry, I misremembered; it was https://github.com/Automattic/wp-calypso/pull/90959 but it was about dev tools and it is now closed 😅

**Comment by arthur791004** [comment]
> The blocker (IMO):
Im noticing some weird behavior on this branch though. When selecting a preview pane page for another site via the ellipsis we get a full reload:

It should be resolved by https://github.com/Automattic/wp-calypso/pull/91034/commits/0fe02fe19479d37b5746a319069c3c165738199f. There was another issue related to the `onSelectionChange`. If we pass to the Dataviews component on Calypso, it might lead to infinite redirection (e.g.: Overview <-> Dev Tools). As a result, I think it would be better to remove it for now.

**Comment by okmttdhr** [comment]
> > The blocker (IMO):
> > Im noticing some weird behavior on this branch though. When selecting a preview pane page for another site via the ellipsis we get a full reload:
> 
> It should be resolved by [0fe02fe](https://github.com/Automattic/wp-calypso/commit/0fe02fe19479d37b5746a319069c3c165738199f). There was another issue related to the `onSelectionChange`. If we pass to the Dataviews component on Calypso, it might lead to infinite redirection (e.g.: Overview <-> Dev Tools). As a result, I think it would be better to remove it for now.

I confirmed this blocker is addressed. 🚀 

https://github.com/Automattic/wp-calypso/assets/5287479/a7293639-e0f6-4882-a3cc-9b63dcbd10ff

Also, I found a potential issue where the sidebar does not sync correctly after selecting an action for an unselected site and then using the browser's back button. Steps to reproduce:

- Click the `...` menu of an unselected site.
- Click `Settings`.
- Use the browser's back button.
- Observe that the first selected site is displayed.

I'm not sure if this needs to be addressed, but it doesn't seem to be a blocker.

https://github.com/Automattic/wp-calypso/assets/5287479/2abedf5f-40a2-4193-9a5b-e57923a038bd

File: client/hosting-overview/index.tsx (hunk 2 above)
Comment (fushar) [reply client/hosting-overview/index.tsx:28]: Wait I'm now a bit confused; in production if I click a site or go to `/overview/` of a fresh new JN site, I get redirected to wp-admin correctly; so what is this change fixing? 🤔

Comment (arthur791004) [reply client/hosting-overview/index.tsx:28]: hmmm...it's redirected by https://github.com/Automattic/wp-calypso/blob/bafc5210661c1bfee22fb1500661b70d2ab82453/client/sites-dashboard-v2/hooks/use-initialize-dataviews-selected-item.ts#L24 😂

Comment (arthur791004) [reply client/hosting-overview/index.tsx:28]: I think it's fine to add  `redirectIfJetpackNonAtomic`. The above one doesn't do the redirection at the beginning, so the user will see the site management panel first and get the redirection later. It might be a little weird to users.

Comment (fushar) [reply client/hosting-overview/index.tsx:28]: I see now... the logic is duplicated here:

https://github.com/Automattic/wp-calypso/blob/10aaf653c41e8c6fc3ffb82a4f956190ee1b9cda/client/sites-dashboard-v2/sites-dataviews/dataviews-fields/site-field.tsx#L83-L84

Would be good if we somehow can consolidate these logic! (Not in this PR)

**Review by fushar — APPROVED** [review]

File: client/hosting-overview/index.tsx (hunk 2 above)
Comment (Addison-Stavlo) [reply client/hosting-overview/index.tsx:28]: Im confused why we are adding this here? Is there a decision around jetpack needing to redirect here?

Design has made other customizations to the overview screens (to copies, etc.) if the site is self-hosted jetpack, so im not entirely sure why we are adding a redirect in this PR?

Comment (Addison-Stavlo) [reply client/hosting-overview/index.tsx:28]: > Wait I'm now a bit confused; in production if I click a site or go to /overview/ of a fresh new JN site

well now im confused as well. things change fast 😵 . last i knew we had updated items in the overview for these sites and didn't redirect them.

Given its already redirecting i guess there is no issue adding this here.

Comment (arthur791004) [reply client/hosting-overview/index.tsx:28]: See https://github.com/Automattic/dotcom-forge/issues/7386. It's already implemented when you select a self-hosted jetpack site

File: client/a8c-for-agencies/components/items-dashboard/items-dataviews/index.tsx [hunk 3]
```diff
@@ -121,6 +121,7 @@ const ItemsDataViews = ( { data, isLoading = false, className }: ItemsDataViewsP
 						return item.id;
 					} )
 				}
+				onSelectionChange={ data.onSelectionChange }
```
Comment (Addison-Stavlo) [inline client/a8c-for-agencies/components/items-dashboard/items-dataviews/index.tsx:124]: I don't see us adding an onSelectionChange on the dotcom side for this?

File: client/a8c-for-agencies/sections/sites/features/jetpack/jetpack-sites-dataviews.tsx [hunk 4]
```diff
@@ -431,6 +435,7 @@ export const JetpackSitesDataViews = ( {
 		actions: [],
 		setDataViewsState: setDataViewsState,
 		dataViewsState: dataViewsState,
+		onSelectionChange: ( [ item ]: SiteData[] ) => openSitePreviewPane( item.site.value ),
```
Comment (Addison-Stavlo) [inline client/a8c-for-agencies/sections/sites/features/jetpack/jetpack-sites-dataviews.tsx:438]: What value is this adding on the a4a side?

**Review by Addison-Stavlo — COMMENTED** [review]
This is testing well for me!

I just had a couple small Qs about the changes on the a4a side and the onSelectionChange.

**Comment by Addison-Stavlo** [comment]
> I'm not sure if this needs to be addressed, but it doesn't seem to be a blocker.

This seems to be in production as well, it seems like a bit of an edge case.

File: client/sites-dashboard-v2/sites-dataviews/index.tsx (hunk 1 above)
Comment (Addison-Stavlo) [reply client/sites-dashboard-v2/sites-dataviews/index.tsx]: Awesome!

I agree this is cleaner than the previous approach or the useEffect. But we may need to do something like that in the future if we want to fix the issue of switching sites with the command palette leaving the selected item in the site list stale. That, another hack, or core changes. 😅

File: client/sites-dashboard-v2/sites-dataviews/dataviews-fields/actions-field.tsx [hunk 5]
```diff
@@ -1,13 +1,28 @@
...
 };
 
 const ActionsField = ( { site }: Props ) => {
+	const handleClickOrKeyDown = ( event: MouseEvent | KeyboardEvent ) => {
+		const target = event.target as HTMLElement;
+
+		// We have to bypass the link to prevent the page from full-reload
+		if ( ! target.closest( 'a' ) ) {
+			event.stopPropagation();
+		}
```
Comment (Addison-Stavlo) [inline client/sites-dashboard-v2/sites-dataviews/dataviews-fields/actions-field.tsx:16]: This was initially confusing me as to why it was resolving the blocker with the hard reload.

If I understand correctly, if we stop propagation inside an anchor our router isn't able to recognize the event and do its work to navigate in the app smoothly. When that happens it defaults to standard anchor behavior of doing a full reload to the href.  Is that what was going on?

File: client/a8c-for-agencies/components/items-dashboard/items-dataviews/index.tsx (hunk 3 above)
Comment (arthur791004) [reply client/a8c-for-agencies/components/items-dashboard/items-dataviews/index.tsx:124]: There is an issue on the dotcom. See https://github.com/Automattic/wp-calypso/pull/91034#issuecomment-2128500688

File: client/a8c-for-agencies/sections/sites/features/jetpack/jetpack-sites-dataviews.tsx (hunk 4 above)
Comment (arthur791004) [reply client/a8c-for-agencies/sections/sites/features/jetpack/jetpack-sites-dataviews.tsx:438]: Adding this can keep the selection item consistent between the Dataviews component and Calypso. We stop the event propagation from the `...`, so the inconsistency might be the edge case. As a result, it's okay not to add this on dotcom side due to the existing infinite redirect issue mentioned above. But for the long time, it would be better to address it and add this on dotcom side as well

File: client/sites-dashboard-v2/sites-dataviews/dataviews-fields/actions-field.tsx (hunk 5 above)
Comment (arthur791004) [reply client/sites-dashboard-v2/sites-dataviews/dataviews-fields/actions-field.tsx:16]: We have the [click event listener](https://github.com/Automattic/wp-calypso/blob/0fe02fe19479d37b5746a319069c3c165738199f/packages/calypso-router/src/index.js#L454) on the root to avoid the full reload.
